Using a SQL Magic Extension: Jupyter Notebook has a built-in extension called "SQL Magic" that allows you to write SQL queries directly in a notebook cell. To use this extension, you need to first install it by running `!pip install ipython-sql` in a notebook cell. Then, you can connect to your SQL database using the following code:

`!conda install -yc conda-forge ipython-sql`

`%load_ext sql`

`%sql postgresql://username:password@host:port/database_name`

In [1]:
%load_ext sql
%sql mysql://root:passwordhere@localhost/projects

'Connected: root@projects'

In [4]:
%sql SELECT * FROM hr LIMIT 5

 * mysql://root:***@localhost/projects
5 rows affected.


[('00-0037846', 'Kimmy', 'Walczynski', '06-04-91', 'Male', 'Hispanic or Latino', 'Engineering', 'Programmer Analyst I', 'Headquarters', '1/20/2002', '', 'Cleveland', 'Ohio'),
 ('00-0041533', 'Ignatius', 'Springett', '6/29/1984', 'Male', 'White', 'Business Development', 'Business Analyst', 'Headquarters', '04-08-19', '', 'Cleveland', 'Ohio'),
 ('00-0045747', 'Corbie', 'Bittlestone', '7/29/1989', 'Male', 'Black or African American', 'Sales', 'Solutions Engineer Manager', 'Headquarters', '10-12-10', '', 'Cleveland', 'Ohio'),
 ('00-0055274', 'Baxy', 'Matton', '9/14/1982', 'Female', 'White', 'Services', 'Service Tech', 'Headquarters', '04-10-05', '', 'Cleveland', 'Ohio'),
 ('00-0076100', 'Terrell', 'Suff', '04-11-94', 'Female', 'Two or More Races', 'Product Management', 'Business Analyst', 'Remote', '9/29/2010', '2029-10-29 06:09:38 UTC', 'Flint', 'Michigan')]

### Rename id column to emp_id

In [34]:
%%sql
ALTER TABLE hr
CHANGE COLUMN id emp_id VARCHAR(20) NULL

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

### Check data types of all columns

In [48]:
%%sql
DESCRIBE hr

 * mysql://root:***@localhost/projects
14 rows affected.


[('emp_id', 'varchar(20)', 'YES', '', None, ''),
 ('first_name', 'text', 'YES', '', None, ''),
 ('last_name', 'text', 'YES', '', None, ''),
 ('birthdate', 'date', 'YES', '', None, ''),
 ('gender', 'text', 'YES', '', None, ''),
 ('race', 'text', 'YES', '', None, ''),
 ('department', 'text', 'YES', '', None, ''),
 ('jobtitle', 'text', 'YES', '', None, ''),
 ('location', 'text', 'YES', '', None, ''),
 ('hire_date', 'date', 'YES', '', None, ''),
 ('termdate', 'date', 'YES', '', None, ''),
 ('location_city', 'text', 'YES', '', None, ''),
 ('location_state', 'text', 'YES', '', None, ''),
 ('age', 'int', 'YES', '', None, '')]

### Change birthdate values to date

**In this code, we first check if the value contains a forward slash '/' using the LIKE operator. If it does, we assume the format is '%m/%d/%Y' and convert the value using the STR_TO_DATE() and DA
TE_FORMAT() functions to the '%Y-%m-%d' format. If it contains a dash '-', we assume the format is '%m-%d-%y' and convert the value to the '%Y-%m-%d' format. If the value does not match either format, we set the birthdate value to NULL.**

**Note that the DATE_FORMAT() function is used to convert the value to the '%Y-%m-%d' format, which is the standard MySQL date format. You can adjust the format string in the DATE_FORMAT() function to match your specific needs if you prefer a different date format.**

In [25]:
%%sql
UPDATE hr
SET birthdate = CASE
  WHEN birthdate LIKE '%/%' THEN DATE_FORMAT(STR_TO_DATE(birthdate, '%m/%d/%Y'), '%Y-%m-%d')
  WHEN birthdate LIKE '%-%' THEN DATE_FORMAT(STR_TO_DATE(birthdate, '%m-%d-%y'), '%Y-%m-%d')
  ELSE NULL
END;

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

In [47]:
%sql SELECT * FROM hr LIMIT 5

 * mysql://root:***@localhost/projects
5 rows affected.


[('00-0037846', 'Kimmy', 'Walczynski', datetime.date(1991, 6, 4), 'Male', 'Hispanic or Latino', 'Engineering', 'Programmer Analyst I', 'Headquarters', datetime.date(2002, 1, 20), None, 'Cleveland', 'Ohio', 31),
 ('00-0041533', 'Ignatius', 'Springett', datetime.date(1984, 6, 29), 'Male', 'White', 'Business Development', 'Business Analyst', 'Headquarters', datetime.date(2019, 4, 8), None, 'Cleveland', 'Ohio', 38),
 ('00-0045747', 'Corbie', 'Bittlestone', datetime.date(1989, 7, 29), 'Male', 'Black or African American', 'Sales', 'Solutions Engineer Manager', 'Headquarters', datetime.date(2010, 10, 12), None, 'Cleveland', 'Ohio', 33),
 ('00-0055274', 'Baxy', 'Matton', datetime.date(1982, 9, 14), 'Female', 'White', 'Services', 'Service Tech', 'Headquarters', datetime.date(2005, 4, 10), None, 'Cleveland', 'Ohio', 40),
 ('00-0076100', 'Terrell', 'Suff', datetime.date(1994, 4, 11), 'Female', 'Two or More Races', 'Product Management', 'Business Analyst', 'Remote', datetime.date(2010, 9, 29), datetime.date(2029, 10, 29), 'Flint', 'Michigan', 28)]

### Change birthdate column datatype

In [28]:
# change data types
%sql ALTER TABLE hr MODIFY COLUMN birthdate DATE

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

In [60]:
%%sql
DESCRIBE hr

 * mysql://root:***@localhost/projects
13 rows affected.


[('emp_id', 'varchar(20)', 'YES', '', None, ''),
 ('first_name', 'text', 'YES', '', None, ''),
 ('last_name', 'text', 'YES', '', None, ''),
 ('birthdate', 'date', 'YES', '', None, ''),
 ('gender', 'text', 'YES', '', None, ''),
 ('race', 'text', 'YES', '', None, ''),
 ('department', 'text', 'YES', '', None, ''),
 ('jobtitle', 'text', 'YES', '', None, ''),
 ('location', 'text', 'YES', '', None, ''),
 ('hire_date', 'date', 'YES', '', None, ''),
 ('termdate', 'text', 'YES', '', None, ''),
 ('location_city', 'text', 'YES', '', None, ''),
 ('location_state', 'text', 'YES', '', None, '')]

### Convert hire_date values to date

In [42]:
%%sql
UPDATE hr
SET hire_date = CASE
  WHEN hire_date LIKE '%/%' THEN DATE_FORMAT(STR_TO_DATE(hire_date, '%m/%d/%Y'), '%Y-%m-%d')
  WHEN hire_date LIKE '%-%' THEN DATE_FORMAT(STR_TO_DATE(hire_date, '%m-%d-%y'), '%Y-%m-%d')
  ELSE NULL
END

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

### Change hire_date column data type

In [44]:
%sql ALTER TABLE hr MODIFY COLUMN hire_date DATE

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

### Convert termdate values to date and remove time

In [54]:
%%sql
UPDATE hr
SET termdate = date(STR_TO_DATE(termdate, '%Y-%m-%d %H:%i:%s UTC'))
WHERE termdate IS NOT NULL AND termdate != ' '

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

In [165]:
%sql SELECT termdate FROM hr LIMIT 10

 * mysql://root:***@localhost/projects
10 rows affected.


[(None,),
 (None,),
 (None,),
 (None,),
 (datetime.date(2029, 10, 29),),
 (None,),
 (None,),
 (None,),
 (datetime.date(2008, 12, 5),),
 (None,)]

### Convert termdate column to date

In [61]:
%sql ALTER TABLE hr MODIFY COLUMN termdate DATE

 * mysql://root:***@localhost/projects
22214 rows affected.


[]

### Add age column

In [85]:
%%sql
ALTER TABLE hr ADD COLUMN age INT;
UPDATE hr SET age = TIMESTAMPDIFF(YEAR, birthdate, CURDATE());

 * mysql://root:***@localhost/projects
0 rows affected.
22214 rows affected.


[]

In [93]:
%%sql
SELECT 
  MIN(age) AS youngest,
  MAX(age) AS oldest
FROM hr

 * mysql://root:***@localhost/projects
1 rows affected.


[(-46, 57)]

In [94]:
%%sql
SELECT COUNT(*) FROM hr WHERE age <18

 * mysql://root:***@localhost/projects
1 rows affected.


[(967,)]

### Check Termdates in the future

In [125]:
%%sql
SELECT COUNT(*)
FROM hr
WHERE termdate > CURDATE()

 * mysql://root:***@localhost/projects
1 rows affected.


[(1599,)]

In [149]:
%%sql
SELECT COUNT(*)
FROM hr
WHERE termdate = '0000-00-00'

 * mysql://root:***@localhost/projects
1 rows affected.


[(18285,)]

## Questions

1. What is the gender breakdown of employees in the company?
2. What is the race/ethnicity breakdown of employees in the company?
3. What is the age distribution of employees in the company?
4. How many employees work at headquarters versus remote locations?
5. What is the average length of employment for employees who have been terminated?
6. How does the gender distribution vary across departments and job titles?
7. What is the distribution of job titles across the company?
8. Which department has the highest turnover rate?
9. What is the distribution of employees across locations by city and state?
10. How has the company's employee count changed over time based on hire and term dates?

In [96]:
%sql SHOW COLUMNS FROM hr

 * mysql://root:***@localhost/projects
14 rows affected.


[('emp_id', 'varchar(20)', 'YES', '', None, ''),
 ('first_name', 'text', 'YES', '', None, ''),
 ('last_name', 'text', 'YES', '', None, ''),
 ('birthdate', 'date', 'YES', '', None, ''),
 ('gender', 'text', 'YES', '', None, ''),
 ('race', 'text', 'YES', '', None, ''),
 ('department', 'text', 'YES', '', None, ''),
 ('jobtitle', 'text', 'YES', '', None, ''),
 ('location', 'text', 'YES', '', None, ''),
 ('hire_date', 'date', 'YES', '', None, ''),
 ('termdate', 'date', 'YES', '', None, ''),
 ('location_city', 'text', 'YES', '', None, ''),
 ('location_state', 'text', 'YES', '', None, ''),
 ('age', 'int', 'YES', '', None, '')]

### 1. What is the gender breakdown of employees in the company?

In [30]:
%%sql
SELECT gender, COUNT(*) AS count
FROM hr
WHERE age >= 18
GROUP BY gender

 * mysql://root:***@localhost/projects
3 rows affected.


[('Male', 10794), ('Female', 9876), ('Non-Conforming', 577)]

### 2. What is the race/ethnicity breakdown of employees in the company?

In [98]:
%%sql
SELECT race, COUNT(*) AS count
FROM hr
WHERE age >= 18
GROUP BY race
ORDER BY count DESC

 * mysql://root:***@localhost/projects
7 rows affected.


[('White', 6057),
 ('Two or More Races', 3508),
 ('Black or African American', 3443),
 ('Asian', 3388),
 ('Hispanic or Latino', 2402),
 ('American Indian or Alaska Native', 1273),
 ('Native Hawaiian or Other Pacific Islander', 1176)]

### 3. What is the age distribution of employees in the company?

In [46]:
%%sql
SELECT 
  MIN(age) AS youngest,
  MAX(age) AS oldest
FROM hr
WHERE age >= 18

 * mysql://root:***@localhost/projects
1 rows affected.


[(20, 57)]

In [104]:
%%sql
SELECT FLOOR(age/10)*10 AS age_group, COUNT(*) AS count
FROM hr
WHERE age >= 18
GROUP BY FLOOR(age/10)*10;

 * mysql://root:***@localhost/projects
4 rows affected.


[(30, 6170), (40, 5897), (20, 5680), (50, 3500)]

In [102]:
%%sql
SELECT 
  CASE 
    WHEN age >= 18 AND age <= 24 THEN '18-24'
    WHEN age >= 25 AND age <= 34 THEN '25-34'
    WHEN age >= 35 AND age <= 44 THEN '35-44'
    WHEN age >= 45 AND age <= 54 THEN '45-54'
    WHEN age >= 55 AND age <= 64 THEN '55-64'
    ELSE '65+' 
  END AS age_group, 
  COUNT(*) AS count
FROM 
  hr
WHERE 
  age >= 18
GROUP BY age_group
ORDER BY age_group;

 * mysql://root:***@localhost/projects
5 rows affected.


[('18-24', 2709),
 ('25-34', 6067),
 ('35-44', 5978),
 ('45-54', 5607),
 ('55-64', 886)]

In [103]:
%%sql
SELECT 
  CASE 
    WHEN age >= 18 AND age <= 24 THEN '18-24'
    WHEN age >= 25 AND age <= 34 THEN '25-34'
    WHEN age >= 35 AND age <= 44 THEN '35-44'
    WHEN age >= 45 AND age <= 54 THEN '45-54'
    WHEN age >= 55 AND age <= 64 THEN '55-64'
    ELSE '65+' 
  END AS age_group, gender,
  COUNT(*) AS count
FROM 
  hr
WHERE 
  age >= 18
GROUP BY age_group, gender
ORDER BY age_group, gender;

 * mysql://root:***@localhost/projects
15 rows affected.


[('18-24', 'Female', 1235),
 ('18-24', 'Male', 1408),
 ('18-24', 'Non-Conforming', 66),
 ('25-34', 'Female', 2854),
 ('25-34', 'Male', 3051),
 ('25-34', 'Non-Conforming', 162),
 ('35-44', 'Female', 2686),
 ('35-44', 'Male', 3134),
 ('35-44', 'Non-Conforming', 158),
 ('45-54', 'Female', 2675),
 ('45-54', 'Male', 2770),
 ('45-54', 'Non-Conforming', 162),
 ('55-64', 'Female', 426),
 ('55-64', 'Male', 431),
 ('55-64', 'Non-Conforming', 29)]

### 4. How many employees work at headquarters versus remote locations?

In [106]:
%%sql
SELECT location, COUNT(*) as count
FROM hr
WHERE age >= 18
GROUP BY location;

 * mysql://root:***@localhost/projects
2 rows affected.


[('Headquarters', 15992), ('Remote', 5255)]

### 5. What is the average length of employment for employees who have been terminated?

In [22]:
%%sql
SELECT ROUND(AVG(DATEDIFF(termdate, hire_date))/365,0) AS avg_length_of_employment
FROM hr
WHERE termdate <> '0000-00-00' AND termdate <= CURDATE() AND age >= 18;

 * mysql://root:***@localhost/projects
1 rows affected.


[(Decimal('7'),)]

In [45]:
%%sql
SELECT ROUND(AVG(DATEDIFF(termdate, hire_date)),0)/365 AS avg_length_of_employment
FROM hr
WHERE termdate <= CURDATE() AND age >= 18;

 * mysql://root:***@localhost/projects
1 rows affected.


[(Decimal('7.4575'),)]

### 6. How does the gender distribution vary across departments?

In [43]:
%%sql 
SELECT department, gender, COUNT(*) as count
FROM hr
WHERE age >= 18
GROUP BY department, gender
ORDER BY department;

 * mysql://root:***@localhost/projects
38 rows affected.


[('Accounting', 'Female', 1462),
 ('Accounting', 'Male', 1640),
 ('Accounting', 'Non-Conforming', 90),
 ('Auditing', 'Female', 23),
 ('Auditing', 'Male', 27),
 ('Business Development', 'Female', 722),
 ('Business Development', 'Male', 798),
 ('Business Development', 'Non-Conforming', 49),
 ('Engineering', 'Female', 2983),
 ('Engineering', 'Male', 3225),
 ('Engineering', 'Non-Conforming', 179),
 ('Human Resources', 'Female', 820),
 ('Human Resources', 'Male', 867),
 ('Human Resources', 'Non-Conforming', 40),
 ('Legal', 'Female', 134),
 ('Legal', 'Male', 158),
 ('Legal', 'Non-Conforming', 7),
 ('Marketing', 'Female', 230),
 ('Marketing', 'Male', 245),
 ('Marketing', 'Non-Conforming', 5),
 ('Product Management', 'Female', 270),
 ('Product Management', 'Male', 338),
 ('Product Management', 'Non-Conforming', 15),
 ('Research and Development', 'Female', 491),
 ('Research and Development', 'Male', 506),
 ('Research and Development', 'Non-Conforming', 35),
 ('Sales', 'Female', 793),
 ('Sales', 'Male', 906),
 ('Sales', 'Non-Conforming', 46),
 ('Services', 'Female', 770),
 ('Services', 'Male', 815),
 ('Services', 'Non-Conforming', 33),
 ('Support', 'Female', 419),
 ('Support', 'Male', 450),
 ('Support', 'Non-Conforming', 34),
 ('Training', 'Female', 759),
 ('Training', 'Male', 819),
 ('Training', 'Non-Conforming', 44)]

### 7. What is the distribution of job titles across the company?

In [24]:
%%sql
SELECT jobtitle, COUNT(*) as count
FROM hr
WHERE age >= 18
GROUP BY jobtitle
ORDER BY jobtitle DESC;

 * mysql://root:***@localhost/projects
185 rows affected.


[('Web Developer IV', 66),
 ('Web Developer III', 69),
 ('Web Developer II', 74),
 ('Web Developer I', 95),
 ('Web Designer IV', 6),
 ('Web Designer III', 13),
 ('Web Designer II', 4),
 ('Web Designer I', 34),
 ('VP Sales', 5),
 ('VP Quality Control', 43),
 ('VP Product Management', 39),
 ('VP of Training and Development', 1),
 ('VP Marketing', 42),
 ('VP Accounting', 76),
 ('Training Manager', 177),
 ('Trainer III', 173),
 ('Trainer II', 175),
 ('Trainer I', 171),
 ('Tax Accountant', 294),
 ('Systems Administrator IV', 84),
 ('Systems Administrator III', 81),
 ('Systems Administrator II', 66),
 ('Systems Administrator I', 359),
 ('Support Staff III', 19),
 ('Support Staff II', 58),
 ('Support Staff', 94),
 ('Structural Engineer', 293),
 ('Structural Analysis Engineer', 45),
 ('Statistician IV', 2),
 ('Statistician III', 4),
 ('Statistician II', 11),
 ('Statistician I', 12),
 ('Staff Scientist', 31),
 ('Staff Accountant IV', 101),
 ('Staff Accountant III', 90),
 ('Staff Accountant II', 77),
 ('Staff Accountant I', 427),
 ('Solutions Engineer Manager', 206),
 ('Solutions Engineer', 204),
 ('Software Test Engineer IV', 84),
 ('Software Test Engineer III', 82),
 ('Software Test Engineer II', 69),
 ('Software Test Engineer I', 80),
 ('Software Engineer IV', 88),
 ('Software Engineer III', 97),
 ('Software Engineer II', 80),
 ('Software Engineer I', 378),
 ('Software Consultant', 327),
 ('Service Tech III', 319),
 ('Service Tech II', 315),
 ('Service Tech', 315),
 ('Service Manager', 319),
 ('Service Coordinator', 313),
 ('Senior Trainer', 176),
 ('Senior Sales Associate', 14),
 ('Senior Recruiter', 193),
 ('Senior Quality Engineer', 148),
 ('Senior Financial Analyst', 57),
 ('Senior Editor', 138),
 ('Senior Developer', 346),
 ('Senior Cost Accountant', 301),
 ('Senior Attorney', 37),
 ('Sales Representative', 5),
 ('Sales Associate', 6),
 ('Research Associate', 201),
 ('Research Assistant IV', 6),
 ('Research Assistant III', 18),
 ('Research Assistant II', 718),
 ('Research Assistant I', 507),
 ('Research Assistant', 39),
 ('Relationshiop Manager', 201),
 ('Recruiting Manager', 22),
 ('Recruiter', 222),
 ('Quality Engineer', 61),
 ('Quality Control Specialist', 131),
 ('Project Manager', 329),
 ('Programmer IV', 61),
 ('Programmer III', 78),
 ('Programmer II', 77),
 ('Programmer I', 84),
 ('Programmer Analyst IV', 72),
 ('Programmer Analyst III', 78),
 ('Programmer Analyst II', 89),
 ('Programmer Analyst I', 86),
 ('Product Engineer', 56),
 ('Pre-Sales Consultant', 209),
 ('Payment Adjustment Coordinator', 23),
 ('Paralegal', 63),
 ('Operator', 210),
 ('Office Assistant IV', 1),
 ('Office Assistant II', 1),
 ('Office Assistant I', 7),
 ('Media Manager II', 47),
 ('Media Manager I', 124),
 ('Mechanical Systems Engineer', 28),
 ('Marketing Manager', 1),
 ('Marketing Assistant', 8),
 ('Librarian', 27),
 ('Legal Assistant', 19),
 ('Junior Trainer', 175),
 ('Internal Auditor', 50),
 ('Information Systems Manager', 34),
 ('Human Resources Manager', 58),
 ('Human Resources Assistant IV', 11),
 ('Human Resources Assistant III', 14),
 ('Human Resources Assistant II', 10),
 ('Human Resources Assistant I', 14),
 ('Human Resources Analyst II', 591),
 ('Human Resources Analyst', 393),
 ('HR Manager', 193),
 ('Help Desk Technician', 322),
 ('Help Desk Operator', 88),
 ('Graphic Designer', 12),
 ('General Manager', 32),
 ('Financial Analyst', 65),
 ('Financial Advisor', 27),
 ('Executive Secretary', 2),
 ('Executive Assistant', 38),
 ('Engineer IV', 1),
 ('Engineer III', 5),
 ('Engineer II', 6),
 ('Engineer I', 9),
 ('Electrical Engineer', 29),
 ('Editor', 88),
 ('Director of Sales', 9),
 ('Developer IV', 71),
 ('Developer III', 81),
 ('Developer II', 64),
 ('Developer I', 77),
 ('Desktop Support Technician', 302),
 ('Design Engineer', 50),
 ('Database Administrator IV', 76),
 ('Database Administrator III', 81),
 ('Database Administrator II', 86),
 ('Database Administrator I', 81),
 ('Data Visualization Specialist', 434),
 ('Data Coordiator', 34

### 8. Which department has the highest turnover rate?

**"Turnover rate" typically refers to the rate at which employees leave a company or department and need to be replaced. It can be calculated as the number of employees who leave over a given time period divided by the average number of employees in the company or department over that same time period.**

In [41]:
%%sql
SELECT department, COUNT(*) as total_count, 
    SUM(CASE WHEN termdate <= CURDATE() AND termdate <> '0000-00-00' THEN 1 ELSE 0 END) as terminated_count, 
    SUM(CASE WHEN termdate = '0000-00-00' THEN 1 ELSE 0 END) as active_count,
    (SUM(CASE WHEN termdate <= CURDATE() THEN 1 ELSE 0 END) / COUNT(*)) as termination_rate
FROM hr
WHERE age >= 18
GROUP BY department
ORDER BY termination_rate DESC;

 * mysql://root:***@localhost/projects
13 rows affected.


[('Marketing', 480, Decimal('43'), Decimal('410'), Decimal('0.9438')),
 ('Training', 1622, Decimal('183'), Decimal('1338'), Decimal('0.9377')),
 ('Human Resources', 1727, Decimal('188'), Decimal('1430'), Decimal('0.9369')),
 ('Services', 1618, Decimal('169'), Decimal('1337'), Decimal('0.9308')),
 ('Accounting', 3192, Decimal('339'), Decimal('2626'), Decimal('0.9289')),
 ('Engineering', 6387, Decimal('673'), Decimal('5259'), Decimal('0.9288')),
 ('Business Development', 1569, Decimal('145'), Decimal('1307'), Decimal('0.9254')),
 ('Product Management', 623, Decimal('64'), Decimal('512'), Decimal('0.9246')),
 ('Sales', 1745, Decimal('183'), Decimal('1426'), Decimal('0.9221')),
 ('Auditing', 50, Decimal('8'), Decimal('38'), Decimal('0.9200')),
 ('Legal', 299, Decimal('37'), Decimal('237'), Decimal('0.9164')),
 ('Support', 903, Decimal('92'), Decimal('732'), Decimal('0.9125')),
 ('Research and Development', 1032, Decimal('111'), Decimal('830'), Decimal('0.9118'))]

### 9. What is the distribution of employees across locations by state?

In [40]:
%%sql
SELECT location_state, COUNT(*) as count
FROM hr
WHERE age >= 18
GROUP BY location_state
ORDER BY count DESC;

 * mysql://root:***@localhost/projects
7 rows affected.


[('Ohio', 17252),
 ('Pennsylvania', 1066),
 ('Illinois', 833),
 ('Indiana', 666),
 ('Michigan', 650),
 ('Kentucky', 417),
 ('Wisconsin', 363)]

### 10. How has the company's employee count changed over time based on hire and term dates?

**This query groups the employees by the year of their hire date and calculates the total number of hires, terminations, and net change (the difference between hires and terminations) for each year. The results are sorted by year in ascending order.**

In [39]:
%%sql
SELECT 
    YEAR(hire_date) AS year, 
    COUNT(*) AS hires, 
    SUM(CASE WHEN termdate <> '0000-00-00' AND termdate <= CURDATE() THEN 1 ELSE 0 END) AS terminations, 
    COUNT(*) - SUM(CASE WHEN termdate <> '0000-00-00' AND termdate <= CURDATE() THEN 1 ELSE 0 END) AS net_change,
    ROUND(((COUNT(*) - SUM(CASE WHEN termdate <> '0000-00-00' AND termdate <= CURDATE() THEN 1 ELSE 0 END)) / COUNT(*) * 100),2) AS net_change_percent
FROM 
    hr
WHERE age >= 18
GROUP BY 
    YEAR(hire_date)
ORDER BY 
    YEAR(hire_date) ASC;

 * mysql://root:***@localhost/projects
21 rows affected.


[(2000, 211, Decimal('26'), Decimal('185'), Decimal('87.68')),
 (2001, 1082, Decimal('197'), Decimal('885'), Decimal('81.79')),
 (2002, 1012, Decimal('160'), Decimal('852'), Decimal('84.19')),
 (2003, 1088, Decimal('187'), Decimal('901'), Decimal('82.81')),
 (2004, 1087, Decimal('188'), Decimal('899'), Decimal('82.70')),
 (2005, 1038, Decimal('165'), Decimal('873'), Decimal('84.10')),
 (2006, 1069, Decimal('165'), Decimal('904'), Decimal('84.57')),
 (2007, 1058, Decimal('132'), Decimal('926'), Decimal('87.52')),
 (2008, 1061, Decimal('132'), Decimal('929'), Decimal('87.56')),
 (2009, 1094, Decimal('137'), Decimal('957'), Decimal('87.48')),
 (2010, 1050, Decimal('115'), Decimal('935'), Decimal('89.05')),
 (2011, 1057, Decimal('105'), Decimal('952'), Decimal('90.07')),
 (2012, 1059, Decimal('98'), Decimal('961'), Decimal('90.75')),
 (2013, 1042, Decimal('84'), Decimal('958'), Decimal('91.94')),
 (2014, 1014, Decimal('76'), Decimal('938'), Decimal('92.50')),
 (2015, 1011, Decimal('66'), Decimal('945'), Decimal('93.47')),
 (2016, 1076, Decimal('60'), Decimal('1016'), Decimal('94.42')),
 (2017, 1043, Decimal('43'), Decimal('1000'), Decimal('95.88')),
 (2018, 1090, Decimal('39'), Decimal('1051'), Decimal('96.42')),
 (2019, 1038, Decimal('41'), Decimal('997'), Decimal('96.05')),
 (2020, 967, Decimal('19'), Decimal('948'), Decimal('98.04'))]

**In this modified query, a subquery is used to first calculate the terminations alias, which is then used in the calculation for the net_change and net_change_percent column in the outer query.**

In [29]:
%%sql
SELECT 
    year, 
    hires, 
    terminations, 
    (hires - terminations) AS net_change,
    ROUND(((hires - terminations) / hires * 100), 2) AS net_change_percent
FROM (
    SELECT 
        YEAR(hire_date) AS year, 
        COUNT(*) AS hires, 
        SUM(CASE WHEN termdate <> '0000-00-00' AND termdate <= CURDATE() THEN 1 ELSE 0 END) AS terminations
    FROM 
        hr
    WHERE age >= 18
    GROUP BY 
        YEAR(hire_date)
) subquery
ORDER BY 
    year ASC;

 * mysql://root:***@localhost/projects
21 rows affected.


[(2000, 211, Decimal('26'), Decimal('185'), Decimal('87.68')),
 (2001, 1082, Decimal('197'), Decimal('885'), Decimal('81.79')),
 (2002, 1012, Decimal('160'), Decimal('852'), Decimal('84.19')),
 (2003, 1088, Decimal('187'), Decimal('901'), Decimal('82.81')),
 (2004, 1087, Decimal('188'), Decimal('899'), Decimal('82.70')),
 (2005, 1038, Decimal('165'), Decimal('873'), Decimal('84.10')),
 (2006, 1069, Decimal('165'), Decimal('904'), Decimal('84.57')),
 (2007, 1058, Decimal('132'), Decimal('926'), Decimal('87.52')),
 (2008, 1061, Decimal('132'), Decimal('929'), Decimal('87.56')),
 (2009, 1094, Decimal('137'), Decimal('957'), Decimal('87.48')),
 (2010, 1050, Decimal('115'), Decimal('935'), Decimal('89.05')),
 (2011, 1057, Decimal('105'), Decimal('952'), Decimal('90.07')),
 (2012, 1059, Decimal('98'), Decimal('961'), Decimal('90.75')),
 (2013, 1042, Decimal('84'), Decimal('958'), Decimal('91.94')),
 (2014, 1014, Decimal('76'), Decimal('938'), Decimal('92.50')),
 (2015, 1011, Decimal('66'), Decimal('945'), Decimal('93.47')),
 (2016, 1076, Decimal('60'), Decimal('1016'), Decimal('94.42')),
 (2017, 1043, Decimal('43'), Decimal('1000'), Decimal('95.88')),
 (2018, 1090, Decimal('39'), Decimal('1051'), Decimal('96.42')),
 (2019, 1038, Decimal('41'), Decimal('997'), Decimal('96.05')),
 (2020, 967, Decimal('19'), Decimal('948'), Decimal('98.04'))]

### 11. What is the tenure distribution for each department?

How long do employees work in each department before they leave or are made to leave?

In [34]:
%%sql
SELECT department, ROUND(AVG(DATEDIFF(CURDATE(), termdate)/365),0) as avg_tenure
FROM hr
WHERE termdate <= CURDATE() AND termdate <> '0000-00-00' AND age >= 18
GROUP BY department

 * mysql://root:***@localhost/projects
13 rows affected.


[('Engineering', Decimal('8')),
 ('Services', Decimal('7')),
 ('Human Resources', Decimal('8')),
 ('Business Development', Decimal('8')),
 ('Sales', Decimal('7')),
 ('Auditing', Decimal('9')),
 ('Training', Decimal('8')),
 ('Accounting', Decimal('8')),
 ('Research and Development', Decimal('8')),
 ('Product Management', Decimal('8')),
 ('Support', Decimal('8')),
 ('Legal', Decimal('9')),
 ('Marketing', Decimal('7'))]

## Summary of Findings

- There are more male employees
- White race is the most dominant while Native Hawaiian and American Indian are the least dominant.
- The youngest employee is 20 years old and the oldest is 57 years old
- 5 age groups were created (18-24, 25-34, 35-44, 45-54, 55-64). A large number of employees were between 25-34 followed by 35-44 while the smallest group was 55-64.
- A large number of employees work at the headquarters versus remotely.
- The average length of employment for terminated employees is around 7 years.
- The gender distribution across departments is fairly balanced but there are generally more male than female employees.
- The Marketing department has the highest turnover rate followed by Training. The least turn over rate are in the Research and development, Support and Legal departments.
- A large number of employees come from the state of Ohio.
- The net change in employees has increased over the years.
- The average tenure for each department is about 8 years with Legal and Auditing having the highest and Services, Sales and Marketing having the lowest.

## Limitations
- Some records had negative ages and these were excluded during querying(967 records). Ages used were 18 years and above.
- Some termdates were far into the future and were not included in the analysis(1599 records). The only term dates used were those less than or equal to the current date.